File: backend/app/schemas/extractions.py
# =============================================================================
# File: extractions.py
# Module/Service: Extraction Service (FR7)
# Layer: Schema
# Purpose: Pydantic request/response models for Extractions API (OpenAPI).
# Responsibilities:
#   - ExtractionCreateRequest
#   - ExtractionResponse (maps ORM result_json → result)
# Dependencies:
#   - Pydantic, app.models.enums
# Public Exports:
#   - ExtractionCreateRequest, ExtractionResponse
# Database/Table: extractions (read mapping only)
# Related Modules: OpenAPI Extraction schema (API wired in Part 5)
# Important Notes:
#   - Exposes source_version_id for FE current-vs-old version UX.
#   - Cost/token fields are internal (not in public OpenAPI response).
# =============================================================================

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ExtractionOutputFormat, ExtractionType


class ExtractionCreateRequest(BaseModel):
    extraction_type: ExtractionType
    output_format: ExtractionOutputFormat = ExtractionOutputFormat.json


class ExtractionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    source_version_id: uuid.UUID
    extraction_type: ExtractionType
    output_format: ExtractionOutputFormat
    result: dict[str, Any] = Field(validation_alias="result_json")
    created_at: datetime

File: frontend/types/extractions.ts
/**
 * =============================================================================
 * File: extractions.ts
 * Module/Service: Extraction Service (Web App)
 * Layer: UI
 * Purpose: TypeScript types for Extractions API matching OpenAPI (FR7).
 * Responsibilities:
 *   - Align Extraction / ExtractionType / OutputFormat / Status with backend
 * Dependencies:
 *   - docs/Enterprise_notebooklm_openapi.yaml Extraction schema
 * Public Exports:
 *   - ExtractionType, ExtractionOutputFormat, ExtractionStatus, Extraction,
 *     ExtractionCreateRequest, TableResultPayload
 * Database/Table: extractions
 * Related Modules: lib/extractions.api, features/extractions/*
 * Important Notes: result is null while processing/failed; structured when completed.
 *   Cost/token fields are not public API.
 * =============================================================================
 */

export type ExtractionType = "table" | "figures" | "entities" | "timeline";

export type ExtractionOutputFormat = "json" | "table";

export type ExtractionStatus = "processing" | "completed" | "failed";

/** Shared table-ready shape for output_format=table (headers + rows). */
export type TableResultPayload = {
  headers: string[];
  rows: Array<Record<string, unknown>>;
};

export type Extraction = {
  id: string;
  document_id: string;
  source_version_id: string;
  extraction_type: ExtractionType;
  output_format: ExtractionOutputFormat;
  status: ExtractionStatus;
  result: Record<string, unknown> | TableResultPayload | null;
  created_at: string;
};

export type ExtractionCreateRequest = {
  extraction_type: ExtractionType;
  output_format?: ExtractionOutputFormat;
};
